scripts/Calc_hru_wgtavg_nc.py

```diff
@@ -89,115 +80,62 @@ def getNetCDFAtt(fn, varname,attName):
-        ij=np.where(IdVal==overlapsId[j])
-        # if nc_in netCDF does not cover nc_wgt domain - case2  
-        if len(ij[0]) == 0:
-          numvoid = numvoid+1
-          wgtArray[j]=0
-        else:
-          a = dataVal[10,ij[0][0],ij[1][0]].tolist()
-          #if value of overlapping polygon is missing data -case1
-          if a == None: 
-            numvoid = numvoid+1
-            wgtArray[j] = 0 
-    
-    #Initialize Val2[ntime] 
-    Val2 = np.zeros((dim1size))
+    a = np.zeros( (dim1size,len(wgtArray)) )
+    for j in range(overlaps):                     # Go through each overlapping polygon
+      # find index of grid cell that match up with hru id of overlapsId[j]
+      ij=np.column_stack(np.where(ghruid==overlapsId[j]))
+      # if nc_in netCDF does not cover nc_wgt domain - case2
+      if len(ij[0]) == 0:
+        numvoid+=1
+        wgtArray[j]=0.0
+      else:
+        a[:,j] = dataVal[:,ij[0,0],ij[0,1]]
+        #if value of overlapping polygon is missing data -case1
+        if any(a[:,j]==FillVal):
+          numvoid+=1
+          wgtArray[j] = 0.0
     if np.nansum(wgtArray) > 0.0:
-    #  print (np.nansum(wgtArray) > 0.0)
       # Adjust weight value if valid weight value (> 0) exist in list
-      newWgtArray = [x/np.nansum(wgtArray) for x in wgtArray]
-    #  print "%d invalid polygons out of %d polygons" %(numvoid,overlaps)
-    #  print "sum of original weights is %f" %np.nansum(wgtArray)
-    #  print "sum of new weights is %f" %np.nansum(newWgtArray)
-    #  print newWgtArray
-      # Go through each overlapping polygon
-      c=0
-      for j in range(len(newWgtArray)): 
-        if overlapsId[j] >= 0: # if there is at least one intersecting polygon
-
-          ij=np.where(IdVal==overlapsId[j]) # ij is tuple e.g., ij=(array[i1],array[i2]) 
-          yj = ij[0]
-          xi = ij[1]
-          Val1=newWgtArray[j]*dataVal[:,yj,xi]
-          #just check If selected hru id from runoff nc matches w/ wieght nc, if not,something wrong exit 
-          if (IdVal[yj,xi] != overlapsId[j]):
-             print "hru_id %d not exist, Sayonara" %IdVal[yj,xi]
-             sys.exit()
-          # Concatenate time series array, Val1 to oriVal containin time serise from all the intersected polygon 
-          if Val1.size: #if Val1 is not empty i.e., exclude empty value in weight list
-            c += 1
-            if c == 1:
-              oriVal = Val1
-            else:
-              oriVal=np.concatenate((oriVal, Val1), axis=1)
-      # This is weighted average value at current hru
-      Val2 = oriVal.sum(axis=1)
-      del oriVal
-
-    #if there is no valid weight values (i.e. all zero)- nothing overlapping for current hru, so missing value
-    else:
-      Val2 = np.ones((dim1size))*-9999.0
-    # store weighted average time series, Val2 in final array wgtVal
-    wgtVal[:,i]= Val2
-    
-    del Val2
+      wgtArray = wgtArray/np.nansum(wgtArray)
+      wgtArray = np.where(np.isnan(wgtArray),0,wgtArray)
```

Comment:
what is the typical shape/rank of wgtArray? It may be faster to use bottlenecks `nansum` here. Also you compute the same value twice `np.nansum(wgtArray)` so save it as a variable.

Reply:
wgtArray is a vector with varying size for different hrus.  `np.nansum(wgtArray)` is saved as `sum_wgt`